### PR TITLE
mon,osd,bluestore: expose object store suboptimal behavior warnings

### PR DESCRIPTION
**Pull request by ifed01**
Now one can see bluefs spillover or lacking compressor plugin warnings via 'ceph health' report.
Sample 'ceph health detail' output:

HEALTH_WARN BlueFS spillover detected at 2 OSD(s); BlueStore compression broken at 1 OSD(s); Module 'restful' has failed dependency: No module named 'OpenSSL'
BLUEFS_SPILLOVER BlueFS spillover detected at 2 OSD(s)
     osd:0 spilled over to slow device: 3 MiB
     osd:1 spilled over to slow device: 3 MiB
BLUESTORE_NO_COMPRESSION BlueStore compression broken at 1 OSD(s)
     osd:2 unable to load:lz4
MGR_MODULE_DEPENDENCY Module 'restful' has failed dependency: No module named 'OpenSSL'

Signed-off-by: Igor Fedotov <ifedotov@suse.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -5380,6 +5383,14 @@ int BlueStore::_balance_bluefs_freespace(PExtentVector *extents)
   bluefs->get_usage(&bluefs_usage);
   ceph_assert(bluefs_usage.size() > bluefs_shared_bdev);
 
+  _clear_alert(ALERT_BLUEFS_SPILLOVER);
+  if (bluefs_shared_bdev == BlueFS::BDEV_SLOW) {
+    auto& p = bluefs_usage[bluefs_shared_bdev];
+    if (p.first != p.second) {
+      _set_alert(ALERT_BLUEFS_SPILLOVER);
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: I'm concerned about this one.  It is totally valid to have an OSD with a big HDD and a small about of SSD space, with maybe only L0 and L1 on it.  It's faster than having no SSD at all.   Spillover isn't a "health" issue, it's a symptom of the cost/performance tradeoff the admin made.  It seems like we should make spillover visible in some other way than a health warning...

Comment (liewegas) [reply src/os/bluestore/BlueStore.cc]: @markhpc @jdurgin @neha-ojha @LenzGr

Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: @liewegas - may be make "spillover" warning configurable? Simple enable/disable switch or spillover size threshold parameter to start warning at?

File: src/os/ObjectStore.h [hunk 2]
```diff
@@ -1550,7 +1550,7 @@ class ObjectStore {
     return false;   // assume a backend cannot, unless it says otherwise
   }
 
-  virtual int statfs(struct store_statfs_t *buf) = 0;
+  virtual int statfs(struct store_statfs_t *buf, osd_alert_list_t& alerts) = 0;
```
Comment (liewegas) [inline src/os/ObjectStore.h]: let's do osd_alert_list_t *alerts since it's an output argument

Comment (liewegas) [reply src/os/ObjectStore.h]: could also make it default to nullptr and make it optional if we want to avoid so much interface churn

**Comment by ifed01** [comment]
@liewegas - all comments but "spillover isn't a warning state' are resolved.

**Comment by liewegas** [comment]
Hmm, so I went to add a patch on top of this to define a mon_warn_on_bluefs_spillover bool option, but it isn't easy because the alerts are just bare strings.

Instead, I'm thinking we should make the osd_alerts_t something more like map<string,list<string>> so that we can define the health alert code (BLUESTORE_BLUEFS_SPILLOVER and BLUESTORE_NO_COMPRESSOR) with associated documentation in doc/rados/opertaions/health-alerts.rst, and can also do the roll-up ("12 OSDs report bluefs spillover" with detail that identifies which OSD(s)) and apply the settings for severity at the mon.

s/osds_alerts_t/osd_alerts_t/, or maybe just health_alert_details_t

**Comment by ifed01** [comment]
@liewegas - IMO the flag 'warn_on_bluefs_spillover' should exist on per-OSD basis. Indeed user can have different OSD configuration. Spillovers might be fine for some of them and aren't for others.
In this case there is probably no need to expose alert code as filtering is performed at OSD level.

Roll-ups are currently performed using osds_alerts_t which is a mapping from alert string to set of OSD ids. At OSD level this set is trivial - just a single id but PGMap maintains osd_sum (osd_stat_t struct) which has osds_alerts_t as a member and builds corresponding mapping in the 'rolled up" form.
So the 'ceph health detail' output looks like:
_HEALTH_WARN Module 'restful' has failed dependency: No module named 'OpenSSL'; Object store may operate suboptimally
MGR_MODULE_DEPENDENCY Module 'restful' has failed dependency: No module named 'OpenSSL'
OBJECT_STORE_WARN Object store may operate suboptimally
    Compressor plugin is not operational. OSDs: 0, 1, 2_

and 'ceph status':

 _cluster:
    id:     89642d14-cd43-44d9-aebd-bd85619314ae
    health: HEALTH_WARN
            Module 'restful' has failed dependency: No module named 'OpenSSL'
            Object store may operate suboptimally_

**Comment by ifed01** [comment]
@liewegas - do you still think we should expose alert codes along with alert text from Obj Store?
So
- Object Store to expose vector<pair<alert_text:string, alert_code:int>>
- OSD to expose osd_id + vector<pair<alert_text:string, alert_code:int>>
- monitor to track OSD alerts in a form of 
  map<pair<alert_text:string, alert_code:int>, osd_ids:set<int>>
and hence we can filter alerts out at monitor level

Or (as per my above comment) do filtering alerts out at object store level
and:
- Object Store to expose vector<alert_text:string>
- OSD to expose osd_id + vector<alert_text:string>
- monitor to track OSD alerts in a form of 
  map<alert_text:string, osd_ids:set<int>>

**Comment by liewegas** [comment]
@ifed01 I'm not sure what alert_code:int is in this case or why it's paired with alert_text:string.  How about:
- ObjectStore: map<string,string> // code->detail, e.g., BLUEFS_SPILLOVER -> "20GiB of bluefs content has spilled over from the 10GiB db volume"
- OSD: same, with "osd.NNN: " prefix on the front of the detail
- monitor: health alert is something like "BLUEFS_SPILLOVER 12 OSD(s) have spillover", with 23 detail strings naming the OSD and other details in the string
- bool mon_warn_on_bluefs_spillover in options.cc

**Comment by ifed01** [comment]
@liewegas - alert_code is a numeric constant from some enum identifying alert type, i.e. the same as code string in your case but numeric.
Your version looks fine except the lack of (well, actually more complicated to implement if needed) ability to filter out spillover warning on per-OSD basis. As I mentioned before spillover might be expected for some OSDs while it's an alert case for others.

**Comment by liewegas** [comment]
You're right, ``osd_warn_on_bluefs_spillover`` is better than ``mon_warn_on_...``.

It seems like it is still better to stick with the detail string, though, instead of a set<int> osds, so that we can include information about the extent of the spillover etc.

**Comment by ifed01** [comment]
@liewegas - updated as per discussed.

File: src/mon/PGMap.cc [hunk 3]
```diff
@@ -2939,6 +2939,50 @@ void PGMap::get_health_checks(
...
+	summary = "BlueFS spillover detected";
+      } else if (asum.first == "BLUESTORE_NO_COMPRESSION") {
+	summary = "BlueStore compression broken";
+      }
+      summary += " at ";
+      summary += stringify(asum.second.first);
+      summary += " OSD(s)";
+      auto& d = checks->add(asum.first, HEALTH_WARN, summary);
+      for (auto& s : asum.second.second) {
+        d.detail.push_back(s);
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/mon/PGMap.cc]: this is performed before when accumulating into os_alerts_sum, see tricks with left/max above.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2019-01-31_22:28:36-rados-wip-sage2-testing-2019-01-30-1657-distro-basic-smithi/

**Pull request merged**